Document remaining IDExxxx rules (#20986)
dotnet/docs: docs/fundamentals/code-analysis/style-rules/unnecessary-code-rules.md

---
title: "Unnecessary code rules"
description: "Learn about code analysis unnecessary code rules"
ms.date: 09/30/2020
ms.topic: reference
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
---
# Unnecessary code rules

> Unnecessary code rules identify different parts of the code base which are unnecessary and hence can be refactored or removed. Presence of unnecessary code indicate one of more of the following problems:

Reviewer: ```suggestion
Unnecessary code rules identify different parts of the code base that are unnecessary and can be refactored or removed. The presence of unnecessary code indicates one of more of the following problems:
```

> - Readability: Code which is unnecessarily degrading readability. For example, [IDE0001](ide0001.md) violations that flag unnecessary type name qualifications.

Reviewer: ```suggestion
- Readability: Code that is unnecessarily degrading readability. For example, [IDE0001](ide0001.md) flags unnecessary type-name qualifications.
```

> - Maintainability: Code which is no longer used after some code refactoring and is unnecessarily being maintained. For example, [IDE0051](ide0051.md) violations that flag unused private fields, properties, events and methods.

Reviewer: ```suggestion
- Maintainability: Code that is no longer used after refactoring and is unnecessarily being maintained. For example, [IDE0051](ide0051.md) flags unused private fields, properties, events, and methods.
```

> - Performance: Unnecessary computation which has no side effects and leads to unnecessary performance overhead. For example, [IDE0059](ide0059.md) violations that flag unused value assignments where the expression to compute value has no side effects.

Reviewer: ```suggestion
- Performance: Unnecessary computation that has no side effects and leads to unnecessary performance overhead. For example, [IDE0059](ide0059.md) flags unused value assignments where the expression to compute a value has no side effects.
```

> - Functionality: Functional issue in code leading to required code being rendered redundant. For example, [IDE0060](ide0060.md) violations that flag unused parameter where the method accidentally ignores an input parameter.

Reviewer: ```suggestion
- Functionality: Functional issue in code leading to required code being rendered redundant. For example, [IDE0060](ide0060.md) flags unused parameters where the method accidentally ignores an input parameter.
```
